beluga/include/beluga/algorithm/unscented_transform.hpp

```diff
@@ -0,0 +1,107 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ALGORITHM_UNSCENTED_TRANSFORM_HPP
+#define BELUGA_ALGORITHM_UNSCENTED_TRANSFORM_HPP
+
+#include <Eigen/Cholesky>
+#include <Eigen/Core>
+
+#include <optional>
+#include <range/v3/range/concepts.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/any_view.hpp>
+#include <range/v3/view/common.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/zip.hpp>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "beluga/eigen_compatibility.hpp"
+namespace beluga {
+
+/**
+ * \brief Implements the unscented transform, a mathematical function used to estimate the result of applying a given
+ *  a possibly nonlinear transformation to a probability distribution that is characterized with mean and covariance.
+ * See https://en.wikipedia.org/wiki/Unscented_transform for more information.
+ *
+ * \tparam MeanDerived Concrete Eigen dense type used for the mean.
+ * \tparam CovarianceDerived Concrete Eigen dense type used for the covariance.
+ * \tparam TransferFn Callable that maps a vector from input space to output space. Possibly non-linear.
+ * \param mean Mean in the input space.
+ * \param covariance Covariance in the input space.
+ * \param transfer_fn Callable so that
+ * \param kappa Parameter used for sigma points selection. See https://arxiv.org/pdf/2104.01958 Equation (5). A sensible
+ * default will be used if not provided.
+ * \return A pair containing (mean, estimate) for the output space.
+ */
+template <typename MeanDerived, typename CovarianceDerived, typename TransferFn>
+auto unscented_transform(
+    const Eigen::MatrixBase<MeanDerived>& mean,
+    const Eigen::MatrixBase<CovarianceDerived>& covariance,
+    TransferFn&& transfer_fn,
+    std::optional<typename MeanDerived::Scalar> kappa = std::nullopt) {
+  using Scalar = typename MeanDerived::Scalar;
+  static_assert(
+      std::is_same_v<typename MeanDerived::Scalar, typename CovarianceDerived::Scalar>,
+      "Mean and covariance scalar types differ.");
+  static_assert(std::is_floating_point_v<Scalar>);
+  static_assert(MeanDerived::ColsAtCompileTime == 1, "Mean should be a column vector.");
+
+  constexpr int kNin = MeanDerived::RowsAtCompileTime;
+  const double k = kappa.value_or(std::max(kNin - 3, 0));
+  assert(k >= 0);
+  const Scalar w0 = static_cast<Scalar>(k) / static_cast<Scalar>(kNin + k);
+  std::vector<Eigen::Vector<Scalar, kNin>> sigma_points;
+  std::vector<Scalar> weights;
+  sigma_points.reserve(2 * kNin + 1);
+  weights.reserve(2 * kNin + 1);
+  sigma_points.emplace_back(mean);
+  weights.emplace_back(w0);
+  const Scalar wn = 1. / (2. * (kNin + k));
+
+  // Compute the Cholesky decomposition of the covariance matrix
+  const Eigen::Matrix<Scalar, kNin, kNin> chol_cov = covariance.llt().matrixL();
```

Comment:
@serraramiro1 consider asserting or throwing on [`covariance.llt().info()`](https://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html#a1d05e5f1a15692959eb9e09d8959a5d5). That'll guard us against non positive semidefinite matrices (ie. broken matrices).